### PR TITLE
If the person inside a morgue tray is clonable, they now beep. Loudly.

### PR DESCRIPTION
**Pull request by Buggy123**
:cl: 
add: Morgue trays now detect if a body inside them possesses a consciousness, and alerts people nearby
/:cl:

Morgue trays will now beep if you put someone in them who can be cloned (and their ghost is in their body).

They will also beep if the person in them re-enters their corpse at any point. Yes, this is spammable. If you keep hearing "BEEP BEEP BEEP BEEP BEEP BEEP BEEP BEEP" maybe you should clone them. Or just open the tray.

This doesn't add any functionality that doesn't already exist - morgues already display a green light if the body is clonable, this only adds a noise to that.

### CODE REVIEW

**Comment by DaxDupont** [comment]
I smash the lights in the morgue so they don't see the tray lights. 
Don't make it beep, that will make it more annoying to hide bodies.

**Comment by yorii** [comment]
Just don't hide your murder victims in the morgue?

**Comment by DaxDupont** [comment]
It's less murder victims, more people we shouldn't clone.

**Comment by RandomMarine** [comment]
How about adding hacking wires to morgue trays to disable sounds/lights?

**Comment by fludd12** [comment]
If you're really bugging out to stop beep spam, just bash their head in. No brain, no clonability.

**Comment by Buggy123** [comment]
That too. Also just don't hide bodies in the morgue. That's probably the second worst place to hide bodies, right after hiding them in genetics.

**Comment by AdamElTablawy** [comment]
>can't mark a body as 'traitor' in a bodybag and stuff it in the morgue anymore

s t o p

**Comment by Ratvar** [comment]
You still can mark a body as "traitor" in a bodybag and stuff it in morgue, n o t h i n g stops you, horray!

**Comment by Jalleo** [comment]
Sheesh it sounds like nobody knows how the cremator works.......

**Comment by Buggy123** [comment]
Or the gibber. Or shuttles.

**Comment by Grazyn** [comment]
This won't change anything. Players already know what morgue lights and corpse description means. "So, why are doctors still leaving my cloneable body in the morgue?" Because those players also rarely play doctors. It's a job mostly played by unexperienced people, and even if it beeps they won't understand what it means. Some coder already tried to fix this in the past by adding a piece of paper in the morgue that explains the lights, and guess what, it didn't work because they don't read it.

If it's not a newb doing it, it means they're doing it on purpose, and in that case adding the beeping only screams "I DED" which isn't a good reason to add a feature.

**Comment by fludd12** [comment]
Actually, Grazyn, I rarely check the morgue because I forget about it when I'm a doctor due to, in the words of a character from The Order of the Stick, `[The] perpetual struggle to maintain the hit points totals of four of five [individuals] determined to deplete them at all costs.` When I'm dead, I don't always feel like staying in my body while it lies rotting in a morgue tray until someone notices the light is green.

**Comment by Xhuis** [comment]
It seems like a good compromise would just be to let you enable or disable the sound with a multitool or wirecutters or some other tool; that way it still works for quick and easy body hiding if people are unattentive, but still serves the function.

**Comment by Buggy123** [comment]
Seems unnecessarily complex TBH, just leave the tray open or do literally anything else with the body. Also not everyone is automatically going to try to spam the sound instead of observing.

**Comment by Incomptinence** [comment]
Yeah it's not the best hiding place but it's hiding the body in plain sight doing stuff like that and your incompetent co-workers don't notice is extremely funny.

**Comment by ChangelingRain** [comment]
>This doesn't add any functionality that doesn't already exist

Why the fuck are you lying? This isn't even like, an believable lie. This is literally adding functionality that doesn't already exist.

Also I hate this, it makes it literally impossible to hide corpses in the morgue, and that's fucking silly. It's a morgue. It's where corpses go.

**Comment by Buggy123** [comment]
Okay, what functionality is it adding that doesn't already exist?

**Comment by DaxDupont** [comment]
The trays now beep for one

**Comment by Iamgoofball** [comment]
@ChangelingRain when was the last time you saw someone hide a corpse in the morgue

**Comment by DaxDupont** [comment]
I did yesterday!

**Comment by Iamgoofball** [comment]
Really? And how do I know you're not just making that up to prove your point?

**Comment by Iamgoofball** [comment]
Because I've played this game for like, 4 years, and never once seen a corpse "hidden" in the morgue when there's 2 ways to destroy corpses acrosss the hallway.

**Comment by DaxDupont** [comment]
That requires you to:
1. Have gloves/luck or emag
2. Drag a corpse to somewhere else than medbay/the morgue which grabs people's attention. Same issue spacing has.

**Comment by Hatfish** [comment]
why is is it a bad thing to have to put risk and effort into permanently removing someone from the round?

**Comment by tortellinitony** [comment]
Yea, this is dumb since you can't use the morgue to hide in plain sight anymore.

**Comment by ChangelingRain** [comment]
Oh yes that's a good point, this will trigger on actual living people trying to be sneaky.
Still a bad idea.

**Comment by DaxDupont** [comment]
**d e s t r o y e d**

**Comment by Buggy123** [comment]
Shit alright i'll try to figure that out.

**Comment by Buggy123** [comment]
Fixed, they will only beep if the body is dead.

**Comment by Buggy123** [comment]
Why won't this compile? This compiles and runs flawlessly locally.

**Comment by Iatots** [comment]
But does the light trigger with living people?

**Comment by Buggy123** [comment]
### Okay, a comprehensive overview:

Current functionality:
If the tray is closed and has at least 1 body, it will use a icon with a red light. The body can be alive or dead.
If the tray is closed and has at least 1 body and at least one of those bodies has a mind in it and it did not commit suicide, it has a green light. Living people trigger this as well as dead people.


new functionality:
If there is a green light (at least one non-suicided body with a mind in it) and that same body is dead, it will beep loudly three times in a row (with literally the most annoying sound I could find, sounds exactly like a alarm clock). This occurs whenever the icon is updated (which, to my understanding, should only be when it is opened or closed or the contents change) and the conditions remain true. That is, it will beep when you close the tray again (but not when you open it), and it will beep if a spirit re-enters a body inside the tray.

So to sum it up:
One or more living person(s) hiding in a tray will have a green light but will not beep.
One or more living person(s) hiding in a tray with any number of dead bodies will have a green light and will not beep.
One or more dead bodies with a spirit in them will beep and have a green light when you close the tray, regardless of living or dead people in the tray.
A ghost re-entering a body inside a tray will turn the light green while the ghost is in their body, and beep every time the ghost re-enters their body(and this absolutely can be spammed, that's the point)


Also, feel like I should also address this: there shouldn't be any fear of ghost communication with this. The beeping takes too long to be a efficient form of communication, and also you can already blink out morse code using the green light more easily than it would be to use the sound. If anything it makes it harder because the constant beeping would alert everyone nearby that something strange is happening. Also if two people are willing to orchestrate communication by morse code like this(no way it could happen accidentally), chances are they'd just metacom instead.

**Comment by KorPhaeron** [comment]
I have no strong feelings on this

**Comment by ChangelingRain** [comment]
"Do three beeps once for yes, three beeps twice for no."

**Comment by KorPhaeron** [comment]
My only hangup with this is that sometimes people hide bodies in the morgue on purpose

**Comment by KorPhaeron** [comment]
Oh okay other people said that already I skipped the comments

**Comment by Buggy123** [comment]
True but, for this to work the ghost has to want to be revived strongly enough to wait for someone to be nearby, and then repeatedly exit-enter their corpse until someone finds them. Or sit in their corpse for who knows how long while that person is hiding them.

I don't think its going to be a frequent occurance, this is more for the incredibly frustrating situation where people walk past your morgue tray that you're in, and because everyone is blind deaf and retarded they _completely ignore_ the little green light and you never get cloned.

**Review by ShizCalev — DISMISSED** [review]

**Comment by ShizCalev** [comment]
Having a wire that you can cut to disable the alarm would be pretty nice.

**Comment by Buggy123** [comment]
I have no idea how to do that, and I don't think the morgue tray can even be screwdrivered so that'd be pretty hard to add.

**Comment by Shadowlight213** [comment]
Just look at disabling security cameras. Its not hard at all.

**Comment by PKPenguin321** [comment]
Needs a set cooldown between beeps that persists even if you exit and re-enter your body to hinder communicating with the dead with this

Perhaps allow alt-clicking (make sure you add an alt-click prompt to the description of the trays if so) or using a multitool on a tray to toggle the speakers

Problems solved

**Comment by PKPenguin321** [comment]
Here I did the toggle part for you
Timers will have to be done yourself though
https://github.com/Buggy123/-tg-station/pull/16

**Comment by Buggy123** [comment]
Alright, gimme a day ish and I'll get it done, I prefer to do code stuff on friday/weekends.

**Comment by optimumtact** [comment]
this is a good idea but you need to make it compile

File: code/game/objects/structures/morgue.dm
Comment (ShizCalev) [inline code/game/objects/structures/morgue.dm]: Move this to the parent.

File: code/game/objects/structures/morgue.dm
Comment (ShizCalev) [inline code/game/objects/structures/morgue.dm]: /obj/structure/bodycontainer/morgue/AltClick

**Review by ShizCalev — CHANGES_REQUESTED** [review]

File: code/game/objects/structures/morgue.dm
Comment (ShizCalev) [inline code/game/objects/structures/morgue.dm]: /obj/structure/bodycontainer/morgue/examine(mob/user)

**Review by ShizCalev — CHANGES_REQUESTED** [review]

**Comment by Buggy123** [comment]
@optimumtact blame PKPenguin321 he offered up some code, except it doesn't actually compile. Definitely don't pay attention to the fact that i merged it without checking it at all first.

Anyway I'll fix it as soon as I have a day where I'm not at college for 12 hours and then sleeping for the other 12 hours.

**Comment by PKPenguin321** [comment]
Haha, my bad. It was a quick job in inbuilt editor. Just fix the paths like shiz said and it'll be fine, needs timers so you can't communicate with ghosts before it should be merged anyways.

**Comment by Buggy123** [comment]
Okay, the beep toggling now works properly and it should compile. I... _can_ add a cooldown but I assure you it isn't useful for preventing ghost communication, you can _already_ do that. I'll make a brief video to prove it.

**Comment by Buggy123** [comment]
[GHOSTCOMMS](http://www.youtube.com/watch?v=pU4XyVTdUFw)

There. I spelled out "SOS" using the **current, unmodified** morgue tray functionality. The beeping does not make it any easier, its much faster to use morse code to communicate than trying to use a long, obvious beeping sound.

**Comment by Buggy123** [comment]
I also complete the changes ShizCalev requested, it just isn't registering it because I fucked up git a little.

**Comment by BeeSting12** [comment]
Beep the tray if soandso is an antag.

BEEP BEEP BEEP BEEP!!!

I hate this PR.

**Comment by Buggy123** [comment]
"Hey ghost, flash the tray green if so and so is a antag"

I'll reiterate, the _current functionality_ is that ghost can turn the morgue tray light green at-will. I must be missing something, please, _please_ explain how this PR makes ghost communication ANY easier for someone who is trying to ghost comm, because I cannot for the life of me see how it does so.

**Comment by PKPenguin321** [comment]
The morse code video he posted has swayed me on this and I think this is fine to merge as is

**Comment by optimumtact** [comment]
can you fix the conflicts please

**Comment by Buggy123** [comment]
Gimme a bit I usually work on shit on friday/weekends.

**Comment by Buggy123** [comment]
Conflicts resolved, ready to merge.

**Comment by ChuckTheSheep** [comment]
Uh, just to note reentering the corpse doesn't update the morgue anymore- that was recently changed so it works more like the cloner in that you can roam around as a ghost. This means it will always be green and eliminates player manipulation. #35072

**Comment by Buggy123** [comment]
Yeah, not sure if I should change that or not.

**Comment by ChuckTheSheep** [comment]
It still will run through the update proc since it's via the logging in/out code, the colors just won't change.

**Pull request merged**
